Add _new/_delete API for ML-KEM/ML-DSA

tamasan238 opened the pull request:
# Description

Added below to solve trouble allocating memory for the MlKemKey structure in C#.
Since the size of the MlKemKey structure depends on the build options, these are needed.

- `wolfCrypt/wc_mlkem.c`: `wc_MlKemKey_New()`, `wc_MlKemKey_Delete()`
- `wolfCrypt/wc_dilithium.c`: `wc_dilithium_new()`, `wc_dilithium_delete()`

This PR is related to another PR I'll add in the comments.

I'm not familiar with adding code to wolfCrypt.
Please feel free to point out any inappropriate parts.

# Testing

Tested.

# Checklist

 - [x] added tests
 - [x] updated/added doxygen
 - [ ] updated appropriate READMEs
 - [ ] Updated manual and documentation

JacobBarthelmeh commented:
@tamasan238 update the "assignees" to be wolfssl-bot when ready for re-review / merge.

tamasan238 commented:
- Added macro guard using `WC_NO_CONSTRUCTORS`
- Removed not-returned `ret`
- (Did not change the documentation style in `dilithium.c`)
  - These are same as other APIs in same file.
  - If changes are necessary, I think it is better to change them all at once in other PR.

I have confirmed that some tests have failed.
I'm investigating the cause, but I don't yet know the reason.

If you have any information, I would appreciate your help.

kojiws commented:
Retest this please

tamasan238 commented:
@kojiws Thank you so much!!
I'll investigate.

tamasan238 commented:
I couldn't reproduce that error in my environment.
I'll retest and see how it goes.

<img width="672" height="530" alt="image" src="https://github.com/user-attachments/assets/e7594953-13b9-4ab5-bdc4-ec79bc7ca1a5" />

Jenkins retest this please

tamasan238 commented:
retest this please

dgarske commented on wolfcrypt/src/wc_mlkem.c:
```diff
@@ -126,6 +126,60 @@ volatile sword16 mlkem_opt_blocker = 0;
+MlKemKey* wc_MlKemKey_New(int type, void* heap, int devId)
+{
+    int ret;
+    MlKemKey* key = (MlKemKey*)XMALLOC(sizeof(MlKemKey), heap,
```
Please add a new dynamic type for DYNAMIC_TYPE_KYBER (or maybe even DYNAMIC_TYPE_MLKEM or DYNAMIC_TYPE_KEM). I'd like this file to use the new type instead of TMP_BUFFER, so please go ahead and refactor it... Or we can do it in a future PR (that would be okay too).

tamasan238 replied:
Thank you for your review!
In order to keep each PR/Commit clear, I would like to implement this in a separate PR.
